70_FLIPFluids/app_resources/compute/diffusion.comp.hlsl

```diff
@@ -0,0 +1,142 @@
+
+    uint3 yn_axiscm = cellIdx.y == gridData.gridSize.y - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 1, 0), gridData.gridSize)].xyz);
+    setYNextMaterial(axisCm, yn_axiscm);
+
+    uint3 zp_axiscm = cellIdx.z == 0 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, -1), gridData.gridSize)].xyz);
+    setZPrevMaterial(axisCm, zp_axiscm);
+
+    uint3 zn_axiscm = cellIdx.z == gridData.gridSize.z - 1 ? (uint3)CM_SOLID : getCellMaterial(axisCellMaterialInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, 1), gridData.gridSize)].xyz);
+    setZNextMaterial(axisCm, zn_axiscm);
+
+    axisCellMaterialOutBuffer[cid].xyz = axisCm;
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void applyDiffusion(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    uint3 axisCm = axisCellMaterialInBuffer[cid].xyz;
+    float3 velocity = (float3)0;
+
+    float3 diff = gridDiffusionInBuffer[cid].xyz;
+    float3 xp_diff = select(isFluidCell(getXPrevMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(-1, 0, 0), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.x * xp_diff;
+
+    float3 xn_diff = select(isFluidCell(getXNextMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(1, 0, 0), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.x * xn_diff;
+
+    float3 yp_diff = select(isFluidCell(getYPrevMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, -1, 0), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.y * yp_diff;
+
+    float3 yn_diff = select(isFluidCell(getYNextMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 1, 0), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.y * yn_diff;
+
+    float3 zp_diff = select(isFluidCell(getZPrevMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, -1), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.z * zp_diff;
+
+    float3 zn_diff = select(isFluidCell(getZNextMaterial(axisCm)),
+        gridDiffusionInBuffer[cellIdxToFlatIdx(cellIdx + int3(0, 0, 1), gridData.gridSize)].xyz, diff);
+    velocity += pc.diffusionParameters.z * zn_diff;
+
+    velocity += pc.diffusionParameters.w * velocityFieldBuffer[cellIdx].xyz;
+
+    enforceBoundaryCondition(velocity, cellMaterialBuffer[cid]);
+
+    gridDiffusionOutBuffer[cid].xyz = velocity;
+}
+
+[numthreads(WorkgroupSize, 1, 1)]
+void updateVelocity(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint cid = ID.x;
+    int3 cellIdx = flatIdxToCellIdx(cid, gridData.gridSize);
+
+    float3 velocity = gridDiffusionInBuffer[cid].xyz;
+
+    enforceBoundaryCondition(velocity, cellMaterialBuffer[cid]);
+
+    velocityFieldBuffer[cellIdx].xyz = velocity;
```

Comment:
I'm pretty sure this can be kernel fused.

Also why do you have so many copies of velocity laying around?
I can count at least 4, do they all need to be live at the same time!?